lib/galaxy/jobs/runners/tes.py

```diff
@@ -0,0 +1,547 @@
+            try:
+                job_id = req.json()["id"]
+                return job_id
+            except KeyError:
+                log.error(f"TES Server fialed to accept the job {req.json()}")
+        except requests.exceptions:
+            log.error(f"{GENERIC_REMOTE_ERROR} on URL {master_addr}")
+
+    def _get_job(self, master_addr: str, job_id: str, view: str = "MINIMAL"):
+        """
+            Get Job Status from TES Server
+        """
+        log.debug(f"Getting status for job id {job_id}")
+        url = f"{master_addr}/v1/tasks/{str(job_id)}"
+        try:
+            req = requests.get(url, params={'view': view})
+            return req.json()
+        except requests.exceptions:
+            log.error(f"{LOST_REMOTE_ERROR} for job id {job_id}")
+
+    def _cancel_job(self, master_addr: str, job_id: str):
+        """
+            Cancel the Job on TES Server
+        """
+        log.debug(f"Cancelling the job id {job_id}")
+        url = f"{master_addr}/v1/tasks/{job_id}:cancel"
+        try:
+            requests.post(url)
+        except requests.exceptions:
+            log.error(f"{GENERIC_REMOTE_ERROR} for cancellation of job id {job_id}")
+
+    def get_upload_path(self, job_id: str, env=None):
+        """
+            For getting upload URL for staging in the files
+        """
+        if env is None:
+            env = []
+
+        encoded_job_id = self.app.security.encode_id(job_id)
+        job_key = self.app.security.encode_id(job_id, kind="jobs_files")
+        endpoint_base = "%sapi/jobs/%s/files?job_key=%s"
+
+        if self.app.config.nginx_upload_job_files_path:
+            endpoint_base = "%s" + \
+                            self.app.config.nginx_upload_job_files_path + \
+                            "?job_id=%s&job_key=%s"
+
+        files_endpoint = endpoint_base % (
+            self.galaxy_url,
+            encoded_job_id,
+            job_key
+        )
+
+        get_client_kwds = dict(
+            job_id=str(job_id),
+            files_endpoint=files_endpoint,
+            env=env
+        )
+        return get_client_kwds
+
+    def find_referenced_subfiles(self, directory: str, command_line: str, extra_files: list):
+        """
+        Return list of files below specified `directory` in job inputs. Could
+        use more sophisticated logic (match quotes to handle spaces, handle
+        subdirectories, etc...).
+        **Parameters**
+        directory : str
+            Full path to directory to search.
+        """
+        if directory is None:
+            return []
+
+        pattern = r'''[\'\"]?(%s%s[^\s\'\"]+)[\'\"]?''' % (escape(directory), escape(os.sep))
+        return self.find_pattern_references(pattern, command_line, extra_files)
+
+    def _read(self, path: str):
+        """
+        Utility method to quickly read small files (config files and tool
+        wrappers) into memory as bytes.
+        """
+        input = open(path, "r", encoding="utf-8")
```

Comment:
You need to open these files as bytes, otherwise it will fail for binary contents.
```suggestion
        input = open(path, "rb", encoding="utf-8")
```